File: src/main/java/org/noxet/noxetserver/minigames/GameDefinition.java
package org.noxet.noxetserver.minigames;

import org.bukkit.GameMode;
import org.noxet.noxetserver.minigames.worldeater.WorldEater;

public enum GameDefinition {
    WORLD_EATER(new MiniGameOptions() {
        @Override
        public String getId() {
            return "world-eater";
        }

        @Override
        public String getDisplayName() {
            return "World Eater";
        }

        @Override
        public int getMinPlayers() {
            return 2;
        }

        @Override
        public int getMaxPlayers() {
            return 8;
        }

        @Override
        public boolean allowPlayerDropIns() {
            return false;
        }

        @Override
        public GameMode getDefaultGameMode() {
            return GameMode.SURVIVAL;
        }

        @Override
        public SpectatorContract getSpectatorContract() {
            return SpectatorContract.ALL;
        }

        @Override
        public int getWorldChunksSquared() {
            return 5;
        }

        @Override
        public MiniGameController initGame() {
            return new WorldEater();
        }

        @Override
        public boolean shouldAnnounceAdvancements() {
            return true;
        }
    });

    private final MiniGameOptions options;

    GameDefinition(MiniGameOptions options) {
        this.options = options;
    }

    public MiniGameOptions getOptions() {
        return options;
    }

    public MiniGameController createGame() {
        return options.initGame();
    }
}

File: src/main/java/org/noxet/noxetserver/minigames/GameDefinition.scala
package org.noxet.noxetserver.minigames

import org.bukkit.GameMode
import org.noxet.noxetserver.minigames.MiniGameOptions.SpectatorContract
import org.noxet.noxetserver.minigames.worldeater

enum GameDefinition(val options: MiniGameOptions):
  case WorldEater extends GameDefinition(new MiniGameOptions {
    override def getId: String = "world-eater"

    override def getDisplayName: String = "World Eater"

    override def getMinPlayers: Int = 2

    override def getMaxPlayers: Int = 8

    override def allowPlayerDropIns: Boolean = false

    override def getDefaultGameMode: GameMode = GameMode.SURVIVAL

    override def getSpectatorContract: SpectatorContract = SpectatorContract.ALL

    override def getWorldChunksSquared: Int = 5

    override def initGame: MiniGameController = worldeater.WorldEater()

    override def shouldAnnounceAdvancements: Boolean = true
  })

  def createGame: MiniGameController = options.initGame